### PR TITLE
API update

### PR DESCRIPTION
**Pull request by sixpearls**
playing with moving tensor API

### CODE REVIEW

**Comment by sixpearls** [comment]
This move introduces some odd calls internally because:
- numpy does not seem as happy to broadcast to trailing axes 
- to keep the appropriate slices of the `__call__` argument C-contiguous for the `_bspl` implementation.

However, it does seem more consistent for the user:

- Everything is zero-indexed, including axis for dimension traversal
- Matches general convention for independent variable axes first. For example, this allows direct interpolation of time series, i.e., from a dataframe `make_interp_spline(df.index, df.values)`

Some user code will be more ugly, such as needing to pass `axis=-1` when stacking to create grids to interpolate or evaluate or selecting output chanels with `[..., idx]`. These aren't too arduous, especially the latter which is not significantly different from selecting a column from a time-series.

File: ndsplines/ndsplines.py [hunk 1]
```diff
@@ -45,41 +45,52 @@ class BSplineNDInterpolator(object):
...
+        
         self.orders = np.broadcast_to(orders, (self.xdim,))
+        self.max_order = np.max(self.orders)
         self.periodic = np.broadcast_to(periodic, (self.xdim,))
         self.extrapolate = np.broadcast_to(extrapolate, (self.xdim,2))
 
-        self.coefficient_op = [0,] + list(i for i in range(2,self.xdim+2)) + [1,]
-        self.u_ops = [[1, i+2] for i in range(self.xdim)]
-        self.output_op = [0,1]
+        assert np.all(coefficients.shape[:self.xdim] == self.xshape - self.orders - 1)
```
Comment (ixjlyons) [inline ndsplines/ndsplines.py]: Isn't it generally favorable to raise an exception? That way you give the user a chance to catch it (probably unlikely in this case) and/or print a more helpful message (i.e. the shape(s) that were expected vs received).

Comment (sixpearls) [reply ndsplines/ndsplines.py]: I was wondering about the convention, I'll fix it.

File: ndsplines/ndsplines.py [hunk 2]
```diff
@@ -91,79 +102,67 @@ def compute_basis_coefficient_selector(self, x, nus=0):
...
-            self.impl.evaluate_spline(t, k, x[i,:], nu, extrapolate_flag, self.interval_workspace[i], self.basis_workspace[i],)
+            self.impl.evaluate_spline(t, k, x[:,i], nu, extrapolate_flag, self.interval_workspace[i], self.basis_workspace[i],)
             np.add(
-                self.coefficient_selector_base[i][..., None],
-                self.interval_workspace[i][:num_points],
-                out=self.coefficient_selector[i, ..., :num_points])
+                self.coefficient_selector_base[i][None, ...],
+                # Broadcasting does not play nciely with xdim as last axis for some reason,
+                # so broadcasting manually. Need to determine speed concequences.
+                self.interval_workspace[i][(slice(0,num_points),) + (None,)*self.xdim],
```
Comment (sixpearls) [inline ndsplines/ndsplines.py]: @ixjlyons Does the azure pipeline run the benchmarks? Or is the best thing for me to run them on my own laptop? I'm curious about how to broadcast this most efficiently.

**Comment by sixpearls** [comment]
It looked like all the doc strings are up to date now.

File: ndsplines/ndsplines.py (hunk 2 above)
Comment (sixpearls) [reply ndsplines/ndsplines.py]: Or if it matters.

Comment (sixpearls) [reply ndsplines/ndsplines.py]: I did a bit of testing and I don't think it will matter too much

Comment (ixjlyons) [reply ndsplines/ndsplines.py]: Just for the record the benchmarks are only run when building the docs but no big deal to add a task to run them in CI as well.

**Comment by sixpearls** [comment]
Makes sense. I wasn't sure if the server would have more variability than
my laptop so I just played with the 1D and 2D (just made it and committed
it to master directly) on my machine. Probably manual benchmarking while
actively developing is sufficient.

>

**Pull request merged**